### DISCUSSION TITLE
Marqo 1.5.1 or 2.2.1?

### DISCUSSION THREAD

**Discussion by heungheung** [DJP-Digital-Jaaduii-Pitara/sakhi-api-service#109] [Q&A]
In AWS Installer
https://github.com/Sunbird-AIAssistant/ai-bot-installer/blob/main/helmcharts/databases/charts/marqo/values.yaml#L8
The marqo tag is 2.2.1

However, in the readme 
https://github.com/Sunbird-AIAssistant/ai-bot-installer/blob/main/README.md
When index documents
```
Download index_documents.py and requirements-dev.txt file from https://github.com/Sunbird-AIAssistant/sakhi-api-service
```
https://github.com/Sunbird-AIAssistant/sakhi-api-service/blob/main/index_documents.py#L112-L123
create index is using index_defaults 

From Marqo doc
https://docs.marqo.ai/2.2.0/API-Reference/Indexes/create_index/
when create index, index_defaults is no longer a param, vs 1.5.1
https://docs.marqo.ai/1.5.1/API-Reference/Indexes/create_index/

Which path shall we go? 
- modify the helm chart to install 1.5.1, or
- modify the payload to use 2.x API?

Side track.... 
1. in requirements-[dev|prod].txt, the marqo lib is 2.1.0, I guess we are talking about using marqo 1.5.x, please note that 1.x API, the pip should be marqo1 while marqo package is now using 2.x API 
2. the requirements pip does NOT work in Python 3.12. I switch to 3.9.x, it works. The container images are using 3.8 and I will assume that should work perfectly

**Comment by heungheung** [comment]
was trying with 1.5.1 but not that simple...... only 2.0+ moved away the privileged mode. 1.5.1 still needs docker privileged mode. In case of Kubernetes 1.5.1, that will be messy...... so, options might be
- use VM + docker, that will be 1.5.1
- use 2.x, then will need to update the API. Not sure if the other will be affected, but at least create index

**Comment by heungheung** [comment]
really got confused after playing around for a few hours........ 

Since marqo is now running in k8s, my first try was going with marqo 2. I also assumed pip in all services will use marqo (i.e. API v2.x instead of marqo1) 

Then I got problems with create index. 
- The first problem is simple, all about parameter names - as the names has been changed in the 2.x API
- 2nd problem is a bit tricky, I got the files indexed - but I hit error when search. After looking into the error log of marqo, I found out search function is not available in unstructured index. By default, index creation with the existing .py will create unstructured index. 
- Then, I recreated the index using structured type. Next, I found I will need to provide the allFields and tensorFields parameters when create the index
- Next is about add document, for structured index, I shouldn't put the actual tensorField when add document. However, the python script throw exception - because the 2.1.0 marqo pip MUST have tensorFields when add document 
- Fine.... update the marqo to 3.0.1, add document seems ok.... how about search? That is something I am not sure yet. 
all the above already took me a few hours.

I shell into the container of activitysakhi, marqo pip is 2.1.0.... that means I am at risk. Will it compatible with marqo 2.x, I am really not sure. 
- Shall I fall back to marqo 1.5.1? 
- If that is the case, does anyone successfully ran 1.5.1 in k8s cluster? 
- Have you seen the docker error? How did you fix it?
Or..... we are talking about we should have a VM for marqo 1.5.1 instead of k8s?

just finished the doc indexing..... did a search and seems activitysakhi can do the search in marqo as well as query openai. But there  is another error which is off topic from this post. For this thread, let's focus 
- what should be the marqo version?
- if 1.5.x, what should be the deployment model (k8s or vm+docker)?
- if 1.5.x and k8s, how to make it work?

(once I have things tested and working, I will shared the required changes for using marqo 2.x)
@kameshbhr

**Reply by heungheung** [reply]
Minor update, as mentioned previously, now using Marqo 2.x and structured index. 

I am not sure I am doing the right thing. For index creation, I use
```
    index_settings = {
            "model": EMBED_MODEL,
            "normalizeEmbeddings": True,
            "type": "structured",
            "textPreprocessing": {
                "splitLength": SPLIT_LENGTH,
                "splitOverlap": SPLIT_OVERLAP,
                "splitMethod": "sentence"
            },
            "allFields": [
                {"name": "text", "type": "text", "features": ["lexical_search"]},
                {"name": "metadata", "type": "text", "features": ["filter"]},
            ],
            "tensorFields": ["text"]
        
    } 
```
when indexing the docs, removed the tensor field
```
      documents=chunk, client_batch_size=_document_batch_size)
```
so far, the sakhi-api-service works fine with the indexed doc in Marqo with the AWS ai bot installer. Able to receive request from telegram unified bot, do search in marqo, send request to openai and provide the response back to telegram. 

Only one minor problem is sakhi-api-service got a typo in the cm which I also put in a different post. Please confirm whether the community should use this Marqo 2.x approach. TIA

**Reply by Jayaprakash8887** [reply]
@heungheung Apologies for the delay in response. We built and tested our application using Marqo 1.5.1. However, we deployed Marqo in a VM instead of K8s. We can deploy our services in K8s and/or vm+docker.

**Reply by heungheung** [reply]
@Jayaprakash8887, I understand we can run/deploy Marqo in either k8s or VM w/docker 

The key points are
- The existing AWS installer will install 2.x https://github.com/Sunbird-AIAssistant/ai-bot-installer/ this allows running Marqo 2.x in k8s
- However, if we want to use Marqo 2.x, the way to create index should be changed, that means the corresponding codes should be updated
- If we don't want to change the existing code for index creation, we will need to use Marqo 1.5.1, where we will need to change the deployment model - that also means we need to update the AWS installer AND will need also need to update all those related codes - because pip for Marqo 1.x has been renamed to marqo1. 
- Things WERE working when the team built, deploy and test - however, we all know that the world change everyday - in our case, the version has been updated, the name of the library (pip) has also been changed. We will also need to update the repos in order to for other adopters can use the codes

**Reply by Jayaprakash8887** [reply]
@heungheung Apologies. I confused your query with respect to other repository. I will let the corresponding team answer your queries. As there is festival in India today and people are on holiday, the team will most likely respond to you tomorrow.

**Reply by heungheung** [reply]
Thanks! No problem. To me, it is not urgent. As I've mentioned in the previous posts, I already have Marqo 2.x and modified the indexing creation. So far, things are working 😄 in my test environment. Just want to see what is the preference of the team and how to move forward.

**Reply by AbhilashKD** [reply]
@heungheung 
please take a latest pull from the repo and then try to run the EKS cluster with the services. I have merged my changes to the repo and fixed the typo error in the marqo version.

**Comment by AbhilashKD** [comment]
@heungheung 
I have checked in both ways so marqo 1.5.1 is working fine in kubernaties cluster and VM+Docker also you can check the VM+Dockerised marqo here 
http://3.111.51.0:8883

And in cluster when you will login to a Pod other than marqo try to curl with URL http://marqo:8883
